### ISSUE TITLE
Network Proxy hangs when large amount of data is transferred

### ISSUE THREAD

**Issue by Jefftree** [kubernetes-sigs/apiserver-network-proxy#76]
This was observed with `kubectl cp` on a large binary file. I attempted to run kubectl cp on a 57M file.

To reproduce, start a kubernetes cluster with network proxy enabled

`KUBE_ENABLE_EGRESS_VIA_KONNECTIVITY_SERVICE=true ./cluster/kube-up.sh`

SSH onto the master and try copying the `kubectl` binary to a random pod:

Eg: 

```
kubectl apply -f https://k8s.io/examples/application/shell-demo.yaml
kubectl cp $(which kubectl) shell-demo:kubectl
```

The `konnectivity-server.log` file shows data transferred, but it stops partially through the data transfer and hangs

EDIT: This can be reproduced without kubernetes

1. Build network proxy with certs & protos generated. (`make build && make gen && make certs`)
1. Copy the kubectl binary (or a similar sized file) to the network proxy dir. (cp $(`which kubectl) .`)
1. Run 4 terminals:

HTTP Connect mode fails:

```
./bin/proxy-server --uds-name=konnectivity-server.socket --mode=http-connect --cluster-cert=certs/agent/issued/proxy-master.crt --cluster-key=certs/agent/private/proxy-master.key --server-port=
./bin/proxy-agent --ca-cert=certs/agent/issued/ca.crt --agent-cert=certs/agent/issued/proxy-agent.crt --agent-key=certs/agent/private/proxy-agent.key
python -m SimpleHTTPServer 8001
./bin/proxy-test-client --proxy-uds=konnectivity-server.socket --proxy-host= --proxy-port=0 --mode=http-connect --request-port=8001 --request-host=localhost --request-path=/kubectl
```

GRPC mode also fails

```
./bin/proxy-server --uds-name=konnectivity-server.socket --mode=grpc --cluster-cert=certs/agent/issued/proxy-master.crt --cluster-key=certs/agent/private/proxy-master.key --server-port=
./bin/proxy-agent --ca-cert=certs/agent/issued/ca.crt --agent-cert=certs/agent/issued/proxy-agent.crt --agent-key=certs/agent/private/proxy-agent.key
python -m SimpleHTTPServer 8001
./bin/proxy-test-client --proxy-uds=konnectivity-server.socket --mode=grpc --proxy-host= --proxy-port=0 --mode=http-connect --request-port=8001 --request-host=localhost --request-path=/kubectl
```

**Comment by caesarxuchao** [comment]
#78 doesn't fix this. The added log shows that the copy stopped after copied ~19MB data, then it stuck at reading from client connection: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/da0186e6db180d4ff153be228cc7465bfe44425f/pkg/agent/agentserver/tunnel.go#L106

**Comment by caesarxuchao** [comment]
#78 doesn't fix the "error dialing backend: EOF" error either.

**Comment by liggitt** [comment]
/reopen

per https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/76#issuecomment-594086871

**Comment by caesarxuchao** [comment]
Per https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/76#issuecomment-594086871, at first I only fixed the race in pendingDials, that didn't fix the issue. Then I fixed another race in calling Send() method of the GRPC stream, that fixed this issue. Sorry I hadn't updated the comment.

I checked our [CI](https://k8s-testgrid.appspot.com/sig-api-machinery-network-proxy#ci-kubernetes-e2e-gci-gce-network-proxy-grpc), there are a few failures but those seem to be https://github.com/kubernetes/kubernetes/issues/86318.

/close
@liggitt let me know if you have other findings.
